src/Renderer/LayeredMaterial.js

```diff
@@ -139,218 +82,209 @@ const LayeredMaterial = function LayeredMaterial(options) {
 };
 
 LayeredMaterial.prototype.setTexturesLayer = function setTexturesLayer(textures, layerType, layer) {
-    const index = this.indexOfColorLayer(layer);
-    const slotOffset = this.getTextureOffsetByLayerIndex(index);
-    for (let i = 0, max = textures.length; i < max; i++) {
-        if (textures[i]) {
-            if (textures[i].texture !== null) {
-                this.setTexture(textures[i].texture, layerType, i + (slotOffset || 0), textures[i].pitch);
-            } else {
-                this.setLayerVisibility(index, false);
-                break;
-            }
-        }
-    }
-};
+    const index = this.indexOfColorLayer(layer.id);
 
-LayeredMaterial.prototype.setTexture = function setTexture(texture, layerType, slot, offsetScale) {
-    if (this.textures[layerType][slot] === undefined || this.textures[layerType][slot].image === undefined) {
-        this.loadedTexturesCount[layerType] += 1;
+    const atlas = _updateAtlas(textures.map(t => t.texture), textures.map(t => t.offsetScale || new THREE.Vector3(0.0, 0.0, 1.0)), layer.transparent);
+
+    atlas.coords = [];
+    for (let i = 0; i < atlas.uv.length; i++) {
+        atlas.coords.push(textures[i].texture.coords);
+        this.offsetScale[l_COLOR][layer.id][i] = atlas.uv[i];
     }
 
-    // BEWARE: array [] -> size: 0; array [10]="wao" -> size: 11
-    this.textures[layerType][slot] = texture || emptyTexture;
-    this.offsetScale[layerType][slot] = offsetScale || new THREE.Vector3(0.0, 0.0, 1.0);
+    atlas._ownerCount = (atlas._ownerCount || 0) + 1;
+    this.uniforms.atlasTextures.value[index] = atlas;
+    this.loadedTexturesCount[l_COLOR] += textures.length;
+
+    this._updateFragmentShader();
 };
 
-LayeredMaterial.prototype.setColorLayerParameters = function setColorLayerParameters(params) {
-    if (this.getColorLayersCount() === 0) {
-        for (let l = 0; l < params.length; l++) {
-            this.pushLayer(params[l]);
-        }
+LayeredMaterial.prototype.setElevationTexture = function setElevationTexture(texture, offsetScale) {
+    if (this.textures[l_ELEVATION][0] === undefined || this.textures[l_ELEVATION][0].image === undefined) {
+        this.loadedTexturesCount[l_ELEVATION] += 1;
     }
+
+    texture._ownerCount = (texture._ownerCount || 0) + 1;
+    // BEWARE: array [] -> size: 0; array [10]="wao" -> size: 11
+    this.textures[l_ELEVATION][0] = texture || emptyTexture;
+    this.offsetScale[l_ELEVATION][0] = offsetScale || new THREE.Vector3(0.0, 0.0, 1.0);
 };
 
 LayeredMaterial.prototype.pushLayer = function pushLayer(param) {
     const newIndex = this.getColorLayersCount();
-    const offset = newIndex === 0 ? 0 : this.getTextureOffsetByLayerIndex(newIndex - 1) + this.getTextureCountByLayerIndex(newIndex - 1);
 
-    this.uniforms.paramLayers.value[newIndex] = new THREE.Vector4();
+    const maxTexturesUnits = Capabilities.getMaxTextureUnitsCount();
+    const nbSamplers = Math.min(maxTexturesUnits - 1, 16 - 1) - 1;
```

Comment:
(this code was there before as well)

I think the reasonning is that sometimes `Capabilities.getMaxTextureUnitsCount();` reported incorrect values, so the `16` was a reasonnable maximum.

This could be investigated and maybe removed...